### PR TITLE
20x speedup for BufferTextWriter

### PR DESCRIPTION
**Pull request by BrannonKing**
I posted this pull request once before, but I've now updated it for 2.2 and fixed the spacing. The JSON serializer uses the TextWriter's `Write(char[], int, int)` overload.

### CODE REVIEW

**Comment by halter73** [comment]
Previous PR: https://github.com/SignalR/SignalR/pull/3378

File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs [hunk 1]
```diff
@@ -200,10 +204,9 @@ internal void OpenConnection(IConnection connection, string data, CancellationTo
 
         public override void LostConnection(IConnection connection)
         {
-            if (_request != null)
-            {
-                _request.Abort();
-            }
+            var request = _request;
+            if (request != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: Keep the style

File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs [hunk 2]
```diff
@@ -107,7 +109,9 @@ internal void OpenConnection(IConnection connection, string data, CancellationTo
                 _stop = true;
 
                 // This will no-op if the request is already finished.
-                ((IRequest)state).Abort();
+                var request = state as IRequest;
+                if (request != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: When would it be null?

File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs [hunk 3]
```diff
@@ -53,7 +53,9 @@ protected override void OnStartFailed()
             // if the transport failed to start we want to stop it silently.
             _stop = true;
 
-            _request.Abort();
+            var request = _request;
+            if (request != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: When would it be null?

File: src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs [hunk 4]
```diff
@@ -438,6 +438,14 @@ public IPerformanceCounter LoadCounter(string categoryName, string counterName,
             // and when they are thrown. 
             try
             {
+                if (!PerformanceCounterCategory.Exists(categoryName))
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs]: We removed the Exists check for a reason. Please remove it.

**Comment by davidfowl** [comment]
This definitely needs unit tests

File: src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs (hunk 4 above)
Comment (BrannonKing) [inline src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs]: I added the check as I have some developers on my team who disable the "just my code" option from time to time. This exception was causing them some grief. I don't believe that having the counters not installed is an unexpected scenario. Help me understand why the check is bad?

File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs (hunk 2 above)
Comment (BrannonKing) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: I don't know when it would be null. I was attempting to fix this issue: https://github.com/SignalR/SignalR/issues/3376 . I was also trying to make the file consistent; I believe that there are null checks for it elsewhere in the file.

File: src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs (hunk 3 above)
Comment (BrannonKing) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: Issue https://github.com/SignalR/SignalR/issues/3376 shows that it can be null, but I don't know when.

Comment (moozzyk) [inline src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs]: This should be a separate PR. The NRE happens if the disconnectToken is tripped before sending the connect request.

**Comment by BrannonKing** [comment]
I've added a unit test for verification of the method behavior, but not verified that the method is used when appropriate. I couldn't figure out a way to accomplish that latter goal. I also added braces to fix the styling.

File: src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs (hunk 4 above)
Comment (rjvdboon) [inline src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs]: The Exists check can be very expensive, look at the Remarks section of the PerformanceCounterCategory.Exists member on msdn. It's much faster to try to open and fail than to check in advance.

**Comment by moozzyk** [comment]
I believe this was ported to 2.2.1
